libdnf/query.cpp

```diff
@@ -0,0 +1,1656 @@
+
+    for (auto match : f.get_matches()) {
+        Id match_evr = pool_str2id(pool, match.str, 1);
+
+        for (Id id = 1; id < pool->nsolvables; ++id) {
+            if (!MAPTST(result, id))
+                continue;
+            Solvable *s = pool_id2solvable(pool, id);
+            int cmp = pool_evrcmp(pool, s->evr, match_evr, EVRCMP_COMPARE);
+
+            if ((cmp > 0 && f.get_cmp_type() & HY_GT) ||
+                (cmp < 0 && f.get_cmp_type() & HY_LT) ||
+                (cmp == 0 && f.get_cmp_type() & HY_EQ)) {
+                MAPSET(m, id);
+            }
+        }
+    }
+}
+
+void
+Query::Impl::filter_nevra(const Filter & f, Map *m)
+{
+    Pool *pool = dnf_sack_get_pool(sack);
+    int fn_flags = (HY_ICASE & f.get_cmp_type()) ? FNM_CASEFOLD : 0;
+
+    for (auto match : f.get_matches()) {
+        const char *nevra_pattern = match.str;
+        if (strpbrk(nevra_pattern, "(/=<> "))
+            continue;
+
+        gboolean present_epoch = strchr(nevra_pattern, ':') != NULL;
+
+        for (Id id = 1; id < pool->nsolvables; ++id) {
+            if (!MAPTST(result, id))
+                continue;
+            Solvable* s = pool_id2solvable(pool, id);
+
+            char* nevra = pool_solvable_epoch_optional_2str(pool, s, present_epoch);
+            if (!(HY_GLOB & f.get_cmp_type())) {
+                if (HY_ICASE & f.get_cmp_type()) {
+                    if (strcasecmp(nevra_pattern, nevra) == 0)
+                        MAPSET(m, id);
+                } else {
+                    if (strcmp(nevra_pattern, nevra) == 0)
+                        MAPSET(m, id);
+                }
+            } else if (fnmatch(nevra_pattern, nevra, fn_flags) == 0) {
+                MAPSET(m, id);
+            }
+        }
+    }
+}
+
+void
+Query::Impl::filter_version(const Filter & f, Map *m)
+{
+    Pool *pool = dnf_sack_get_pool(sack);
+    int cmp_type = f.get_cmp_type();
+
+    for (auto match_in : f.get_matches()) {
+        const char *match = match_in.str;
+        char *filter_vr = solv_dupjoin(match, "-0", NULL);
+
+        for (Id id = 1; id < pool->nsolvables; ++id) {
+            if (!MAPTST(result, id))
+                continue;
+            char *e, *v, *r;
+            Solvable *s = pool_id2solvable(pool, id);
+            if (s->evr == ID_EMPTY)
+                continue;
+            const char *evr = pool_id2str(pool, s->evr);
+
+            pool_split_evr(pool, evr, &e, &v, &r);
+
+            if (cmp_type == HY_GLOB) {
+                if (fnmatch(match, v, 0))
+                    continue;
+                MAPSET(m, id);
+            }
+
+            char *vr = pool_tmpjoin(pool, v, "-0", NULL);
+            int cmp = pool_evrcmp_str(pool, vr, filter_vr, EVRCMP_COMPARE);
+            if ((cmp > 0 && cmp_type & HY_GT) ||
+                (cmp < 0 && cmp_type & HY_LT) ||
+                (cmp == 0 && cmp_type & HY_EQ)) {
+                MAPSET(m, id);
+            }
+        }
+        g_free(filter_vr);
```

Comment:
Why not use `g_autofree` above instead?

Reply:
Do we have `g_autofree` in EL 7? IIRC, this is supposed to still work on EL7...

Reply:
Here the memory was allocated by solv_dupjoin(), teherefore it has to be removed by solv_free().